data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/MilestoneDependentTypesUtil.java

```diff
@@ -79,38 +79,56 @@ SerializableOneOfTypeDefinition<T> getMultipleSchemaDefinitionFromMilestone(
     return builder.build();
   }
 
-  public static <T extends SszData> DeserializableTypeDefinition<? extends T> slotBasedSelector(
-      final String json,
-      final SchemaDefinitionCache schemaDefinitionCache,
-      final Function<SchemaDefinitions, SszSchema<? extends T>> getSchema) {
-    final Optional<UInt64> slot =
-        // SignedBeaconBlock
-        getSlot(json, "message", "slot")
-            // SignedBlockContents
-            .or(() -> getSlot(json, "signed_block", "message", "slot"));
-    final SpecMilestone milestone =
-        schemaDefinitionCache.milestoneAtSlot(
-            slot.orElseThrow(() -> new BadRequestException("Could not locate slot in JSON data")));
-    return getSchema
-        .apply(schemaDefinitionCache.getSchemaDefinition(milestone))
-        .getJsonTypeDefinition();
+  public static <T extends SszData>
+      DeserializableTypeDefinition<? extends T> headerBasedSelectorWithSlotFallback(
+          final Map<String, String> headers,
+          final String json,
+          final SchemaDefinitionCache schemaDefinitionCache,
+          final Function<SchemaDefinitions, SszSchema<? extends T>> getSchema) {
+    if (headers.containsKey(HEADER_CONSENSUS_VERSION)) {
+      return headerBasedSelector(headers, schemaDefinitionCache, getSchema);
+    }
+    return slotBasedSelector(json, schemaDefinitionCache, getSchema);
   }
 
   public static <T extends SszData> DeserializableTypeDefinition<? extends T> headerBasedSelector(
       final Map<String, String> headers,
       final SchemaDefinitionCache schemaDefinitionCache,
       final Function<SchemaDefinitions, SszSchema<? extends T>> getSchema) {
+    if (!headers.containsKey(HEADER_CONSENSUS_VERSION)) {
+      throw new BadRequestException(
+          String.format("Missing required header value for (%s)", HEADER_CONSENSUS_VERSION));
+    }
     try {
       final SpecMilestone milestone = SpecMilestone.forName(headers.get(HEADER_CONSENSUS_VERSION));
       return getSchema
           .apply(schemaDefinitionCache.getSchemaDefinition(milestone))
           .getJsonTypeDefinition();
     } catch (Exception e) {
       throw new BadRequestException(
-          String.format("(%s) header value was unexpected", HEADER_CONSENSUS_VERSION));
+          String.format(
+              "Invalid value for (%s) header: %s",
+              HEADER_CONSENSUS_VERSION, headers.get(HEADER_CONSENSUS_VERSION)));
     }
   }
 
+  private static <T extends SszData> DeserializableTypeDefinition<? extends T> slotBasedSelector(
+      final String json,
+      final SchemaDefinitionCache schemaDefinitionCache,
+      final Function<SchemaDefinitions, SszSchema<? extends T>> getSchema) {
+    final Optional<UInt64> slot =
+        // SignedBeaconBlock
+        getSlot(json, "message", "slot")
+            // SignedBlockContents
+            .or(() -> getSlot(json, "signed_block", "message", "slot"));
```

Comment:
is block contents still a thing? I thought that went away?

Reply:
Yep, we added them since Deneb to handle block and blobs but you might be right since for the new APIs handling block contents we have the consensus version in the header and hence we never fallback to the slot selector. So this can me removed since never used actually